frontend/packages/console-plugin-sdk/src/subscribeToExtensions.ts

```diff
@@ -0,0 +1,141 @@
+import { Store } from 'redux';
+import * as _ from 'lodash';
+import { RootState } from '@console/internal/redux';
+import { isExtensionInUse, PluginStore } from './store';
+import { Extension, ExtensionTypeGuard } from './typings';
+
+let subscriptionServiceInitialized = false;
+
+let onSubscriptionAdded: (sub: ExtensionSubscription<Extension>) => void;
+
+const subscriptions: ExtensionSubscription<Extension>[] = [];
+
+export const initSubscriptionService = (pluginStore: PluginStore, reduxStore: Store<RootState>) => {
+  if (subscriptionServiceInitialized) {
+    throw new Error('Subscription service is already initialized');
+  }
+
+  subscriptionServiceInitialized = true;
+
+  const getAllExtensions = () => pluginStore.getAllExtensions();
+  const getAllFlags = () => reduxStore.getState().FLAGS.toObject();
+
+  type FeatureFlags = ReturnType<typeof getAllFlags>;
+
+  const invokeListener = (
+    sub: ExtensionSubscription<Extension>,
+    currentExtensions: Extension[],
+    currentFlags: FeatureFlags,
+  ) => {
+    // Narrow extensions according to type guards
+    const matchedExtensions = _.flatMap(sub.typeGuards.map((tg) => currentExtensions.filter(tg)));
+
+    // Gate matched extensions by relevant feature flags
+    const extensionsInUse = matchedExtensions.filter((e) => isExtensionInUse(e, currentFlags));
+
+    // Invoke listener only if the extension list has changed
+    if (!_.isEqual(extensionsInUse, sub.listenerLastArgs)) {
+      sub.listenerLastArgs = extensionsInUse;
+      sub.listener(extensionsInUse);
+    }
+  };
+
+  onSubscriptionAdded = (sub) => {
+    invokeListener(sub, getAllExtensions(), getAllFlags());
+  };
+
+  let lastExtensions: Extension[] = [];
+  let lastFlags: FeatureFlags = {};
+
+  const invokeAllListeners = () => {
+    const nextExtensions = getAllExtensions();
+    const nextFlags = getAllFlags();
+
+    if (_.isEqual(nextExtensions, lastExtensions) && _.isEqual(nextFlags, lastFlags)) {
```

Comment:
if you do not covert the Flags ImmutableMap to Object you wont have to use isEqual, just `nextFlags === lastFlags` would be sufficient.
I think its worth doing since `invokeAllListeners` is invoked on any redux state change - instead of calling `toObject` and `isEqual` you would need to do just `===` . And once you get inside the `if` block you can convert flags to object

Reply:
Agreed and done :smiley: